File: log.swift
/**************************************************************************/
/*                                                                        */
/*  Copyright (C) 2004-2017 INTERSEC SA                                   */
/*                                                                        */
/*  Should you receive a copy of this source code, you must check you     */
/*  have a proper, written authorization of INTERSEC to hold it. If you   */
/*  don't have such an authorization, you must DELETE all source code     */
/*  files in your possession, and inform INTERSEC of the fact you obtain  */
/*  these files. Should you not comply to these terms, you can be         */
/*  prosecuted in the extent permitted by applicable law.                 */
/*                                                                        */
/**************************************************************************/

import Glibc

extension Logger : Wipeable {
    /// Build a new static logger.
    ///
    /// Build a new logger whose lifetime is the one of the program.
    ///
    /// - Parameter name: the name of the logger. That name will be concatenated
    ///  with `parent`'s name to build the fullname of the logger.
    /// - Parameter parent: the parent of the logger.
    /// - Parameter level: default logging level of the logger.
    public init(name: StaticString,
                parent: UnsafeMutablePointer<Logger>? = nil,
                level: core.LogLevel = LOG_LEVEL_INHERITS)
    {
        self.init()
        assert (name.hasPointerRepresentation)

        self.is_static = true
        self.parent = parent
        self.name = name.utf8Start.withMemoryRebound(to: Int8.self, capacity: name.utf8CodeUnitCount) {
            (ptr: UnsafePointer<Int8>) -> LString in

            let len = strlen(ptr)
            return LString(ptr, count: Int32(len), flags: 0)
        }
        self.level = LOG_UNDEFINED
        self.defined_level = LOG_UNDEFINED
        self.default_level = Int32(level.rawValue)
    }

    mutating func format<C: Collection>(items: C, separator: String) -> String {
        var out = ""

        for item in items {
            if !out.isEmpty && !separator.isEmpty {
                out.write(separator)
            }
            print(item, terminator: "", to: &out)
        }
        return out
    }

    mutating func log<C: Collection>(level: Int32, items: C, separator: String, file: String, function: String, line: Int) {
        _ = self.format(items: items, separator: separator).withLString {
            withVaList([$0.len, $0.data!]) {
                logger_vlog(&self, level, nil, -1, file, function, Int32(line), "%*pM", $0)
            }
        }
    }

    mutating func log<C: Collection>(level: core.LogLevel, items: C, separator: String, file: String, function: String, line: Int) {
        let level = Int32(level.rawValue)

        if !logger_has_level(&self, level) {
            return
        }

        self.log(level: level, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit a log line with the specified criticity `level`
    public mutating func log(level: core.LogLevel, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        self.log(level: level, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit an error log.
    public mutating func error(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        self.log(level: LOG_LEVEL_ERR, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit a warning log.
    public mutating func warning(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        self.log(level: LOG_LEVEL_WARNING, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit a notice log.
    public mutating func notice(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        self.log(level: LOG_LEVEL_NOTICE, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit an informational log.
    public mutating func info(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        self.log(level: LOG_LEVEL_INFO, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit a debug log.
    public mutating func debug(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        self.log(level: LOG_LEVEL_DEBUG, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit a trace line.
    ///
    /// Traces are defined by increasing `level`: the greater the level
    /// is, the more specific the trace is.
    public mutating func trace(level: Int32, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) {
        let logLevel = level + Int32(LOG_LEVEL_TRACE.rawValue)

        if !logger_has_level(&self, logLevel) && __logger_is_traced(&self, level, file, function, self.full_name.data) <= 0 {
            return
        }

        self.log(level: logLevel, items: items, separator: separator, file: file, function: function, line: line)
    }

    /// Emit a log and cleaning exit the program.
    public mutating func exit(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) -> Never {
        return self.format(items: items, separator: separator).withCString {
            return withVaList([$0]) {
                __logger_vexit(&self, file, function, Int32(line), "%s", $0)
            }
        }
    }

    /// Emit a log and terminate the execution of the program.
    public mutating func fatal(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) -> Never {
        return self.format(items: items, separator: separator).withCString {
            return withVaList([$0]) {
                __logger_vfatal(&self, file, function, Int32(line), "%s", $0)
            }
        }
    }

    /// Emit a log a abort the execution of the program.
    public mutating func panic(_ items: Any..., separator: String = " ", file: String = #file, function: String = #function, line: Int = #line) -> Never {
        return self.format(items: items, separator: separator).withCString {
            return withVaList([$0]) {
                __logger_vpanic(&self, file, function, Int32(line), "%s", $0)
            }
        }
    }
}

File: log.py
##########################################################################
#                                                                        #
#  Copyright (C) INTERSEC SA                                             #
#                                                                        #
#  Should you receive a copy of this source code, you must check you     #
#  have a proper, written authorization of INTERSEC to hold it. If you   #
#  don't have such an authorization, you must DELETE all source code     #
#  files in your possession, and inform INTERSEC of the fact you obtain  #
#  these files. Should you not comply to these terms, you can be         #
#  prosecuted in the extent permitted by applicable law.                 #
#                                                                        #
##########################################################################

import logging
import common

from logging import ERROR, WARNING, WARN, INFO, DEBUG, CRITICAL

NOTICE = 25
PANIC = 60

logging.addLevelName("PANIC", PANIC)
logging.addLevelName("NOTICE", NOTICE)

class IntersecLogger(logging.Logger):

    def __init__(self):
        # The python 2.6 implementation of the logging module use the old
        # python class model, as a consequence we must not use super() here.
        logging.Logger.__init__(self, "intersec", 0)

    def panic(self, msg, *args, **kwargs):
        self.log(PANIC, msg, *args, **kwargs)

    def notice(self, msg, *args, **kwargs):
        self.log(NOTICE, msg, *args, **kwargs)

class IntersecLogHandler(logging.Handler):
    '''subclass logging.Handler:
         redefine the emit method by calling python binding of common e_log
         add two levels
       => in this way, when you use a logger with this handle:
          "LOGGER.error" called lib-common "e_error" method etc...
    '''
    level_map = {
        DEBUG:    common.LOG_DEBUG,
        INFO:     common.LOG_INFO,
        NOTICE:   common.LOG_NOTICE,
        WARNING:  common.LOG_WARNING,
        WARN:     common.LOG_WARNING,
        ERROR:    common.LOG_ERR,
        CRITICAL: common.LOG_CRIT,
        PANIC:    common.LOG_PANIC,
    }

    def emit(self, record):
        level = self.level_map.get(record.levelno, common.LOG_ERR)
        msg = self.format(record)
        common.log(level, msg)

MYHANDLER = IntersecLogHandler()
LOGGER = IntersecLogger()
LOGGER.addHandler(MYHANDLER)

#example
if __name__ == "__main__":
    TMP = "arg"
    LOGGER.error("I m an error log")
    LOGGER.warning("I m not an error log, I m a warning log")
    LOGGER.warn("mee too :D")
    LOGGER.warn("And I could be use with some: %s", TMP)
    LOGGER.notice("I m noticing you that...")
    LOGGER.debug("I m debugging")
    LOGGER.info("info")
    LOGGER.panic("time to panic and generate a core")
